questionpy_server/worker/worker/__init__.py

```diff
@@ -75,6 +75,14 @@ def send(self, message: MessageToWorker) -> None:
     async def get_resource_usage(self) -> WorkerResources | None:
         """Get the worker's current resource usage. If unknown or unsupported, return None."""
 
+    @abstractmethod
+    async def load_package(self, *, reload: bool = False) -> None:
+        """Loads a package.
+
+        Args:
+            reload: Force reload of module.
+        """
```

Comment:
Wollen wir das vielleicht noch etwas konkretisieren?
```suggestion
    @abstractmethod
    async def reload_main_package(self) -> None:
        """Reloads the main package."""
```

Der Worker lädt das Main-Paket ja erstmalig von alleine, und andere Pakete können damit noch nicht geladen werden.

Reply:
> Der Worker lädt das Main-Paket ja erstmalig von alleine

Was meinst du mit "von alleine". Der Worker lädt das Paket in `BaseWorker._initialize()`. Dort wird auch `load_package()` aufgerufen. Daher `load_package` und nicht `reload_package`.

> und andere Pakete können damit noch nicht geladen werden.

Die Pakete werden in dem dict `loaded_packages` schon überschrieben:

```py
        self.loaded_packages[str(msg.location)] = package
````

Allerdings nicht die Einträge aus `sys.modules` entfernt...

Also, die aktuelle Implementierung basiert auf den Annahmen, dass

1. sich die Namen der Paket-Namespaces und -Shortnames nicht ändern. Ansonsten müsste man `package.setup_imports()` vor dem Neuladen auch rückgängig machen (also die Pfade aus `sys.path` löschen).

1. die "anderen" Pakete sich nicht geändert haben, sonst müsste man die auch aus `sys.modules` entfernen. (Ist für den konkreten Use case (watch mode) auch nicht relevant, da man ja immer am "main" package schraubt.)

Also, dann eher **`load_main_package`** nennen?

(siehe auch Kommentar für mehr Kontext/alternative Lösung.)

Reply:
> Was meinst du mit "von alleine"

Im `start()`-Call, ohne weitere Einwirkung von außen.

> Der Worker lädt das Paket in BaseWorker._initialize(). Dort wird auch load_package() aufgerufen. Daher load_package und nicht reload_package.

Mir ging es hier jetzt um die public API, ein `load_main_package` macht nur intern Sinn. Von außen wird das Main-Paket ausschließlich <code><b>re</b>load</code>ed werden.